### PR TITLE
Implementation of multiple periodic boundary conditions in DEM

### PR DESCRIPTION
**Pull request by blaisb**
### Description

This PR enables multiple pairs of periodic boundaries for the DEM module, whereas before only one pair of periodic boundaries at a time was supported. This allows particles to traverse periodic corners of the domain at vertices where many periodic boundaries meet.

Most of this PR involves refactoring scalar quantities identifying attributes of a single periodic boundary pair to ``std::vector`` or ``std::unordered_map`` quantities for multiple periodic boundary pairs.

This is a recreation of #1949 since that branch had received merge commits.

### Testing

six new application tests for ``lethe-particles`` were added, simulating particle collisions through corners and faces. An additional unit test was added also.

### Documentation

The Boundary Conditions section for the DEM Parameters Guide has been updated to provide more details about the parameters for periodic boundaries, and now mentions support for multiple periodicity.

### Checklist (will be removed when merged)
See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#pull-requests) for more information about the pull request process.

Code related list:
- [x] All in-code documentation related to this PR is up to date (Doxygen format)
- [x] Copyright headers are present and up to date
- [x] Lethe documentation is up to date
- [x] New feature has unit test(s) (preferred) or application test(s), and restart files are in the generator folder
- [x] The branch is rebased onto master
- [ ] Changelog (CHANGELOG.md) is up to date
- [x] Code is indented with indent-all and .prm files (examples and tests) with prm-indent

Pull request related list:
- [x] Labels are applied
- [x] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [x] If this PR closes an issue or is related to a project, it is linked in the "Projects" or "Development" section
- [x] If the fix is temporary, an issue is opened
- [x] The PR description is cleaned and ready for merge

### CODE REVIEW

File: source/core/parameters_lagrangian.cc [hunk 1]
```diff
@@ -1701,11 +1707,15 @@ namespace Parameters
           boundary_rot_speed.insert({counter, 0});
           boundary_rot_vector.insert({counter, zero_tensor});
           point_on_rot_axis.insert({counter, Point<3>(zero_tensor)});
+          // periodic_bndry_0.insert({counter, 0});
+          // periodic_bndry_1.insert({counter, 0});
+          // periodic_dir.insert({counter, 0});
```
Comment (karlwilson) [inline source/core/parameters_lagrangian.cc]: I think this can this be removed

**Review by karlwilson — COMMENTED** [review]
Everything looks good

File: doc/source/parameters/dem/model_parameters.rst [hunk 2]
```diff
@@ -174,13 +174,14 @@ See how the mechanism works with mobility status in the figure below:
...
 * ``granular temperature threshold`` is the threshold of the granular temperature below which the contacts are disabled.
 * ``solid fraction threshold`` is the minimum solid fraction of the cell in which the contacts may be disabled.
 
 Some parameters in the load balance section may be used to improve the performance of the dynamic disabling contacts feature using the dynamic load balancing.
+
 .. note::
-The ``load balance method`` may be set to ``dynamic_with_sparse_contacts`` and factors of the weight of the cells by mobility status may be adjusted using the ``active weight factor`` and ``inactive weight factor`` parameters. There is factor only for active and inactive status, mobile factor is always 1.
+    The ``load balance method`` may be set to ``dynamic_with_sparse_contacts`` and factors of the weight of the cells by mobility status may be adjusted using the ``active weight factor`` and ``inactive weight factor`` parameters. There is factor only for active and inactive status, mobile factor is always 1.
```
Comment (jatred) [inline doc/source/parameters/dem/model_parameters.rst:184]: Why are the parameters ``active weight factor`` and ``inactive weight factor`` mentioned in a note here and not in the code-block like the rest?

We could do ``set active weight factor = x    # if method = dynamic_with_sparse_contacts`` or something like that if I understand. It would allow us to list them explicitly like for ``step`` or ``frequency`` and make it a bit clearer in my opinion.

File: doc/source/parameters/dem/insertion_info.rst [hunk 3]
```diff
@@ -137,7 +137,7 @@ The ``plane`` insertion method inserts particles at the centroid of insertion ce
 --------------------
 List
 --------------------
-The ``list`` insertion method insert particles at precis coordinates with specific velocities (translational and angular) and diameters.  This method is preferred for small number of particles.
+The ``list`` insertion method insert particles at precise coordinates with specific velocities (translational and angular) and diameters.  This method is preferred for small number of particles.
```
Comment (jatred) [inline doc/source/parameters/dem/insertion_info.rst]:
```suggestion
The ``list`` insertion method inserts particles at precise coordinates with specific velocities (translational and angular) and diameters.  This method is preferred for small numbers of particles.
```

**Review by jatred — COMMENTED** [review]
I've only reviewed the tests and .rst files, I will continue with the code later

**Review by jatred — APPROVED** [review]
I couldn't find anything that seemed off in the code, it looks really good with all the changes. The @briefs also look complete, no parameter missing from what I've seen. Nice!

File: source/dem/periodic_boundaries_manipulator.cc [hunk 4]
```diff
@@ -178,6 +231,43 @@ PeriodicBoundariesManipulator<dim>::check_and_move_particles(
...
 
+template <int dim>
+void
+PeriodicBoundariesManipulator<dim>::compute_combined_periodic_offsets()
+{
+  this->combined_periodic_offsets.clear();
+
+  for (auto const &[id, offset] : this->periodic_offsets)
+    {
+      size_t current_size = this->combined_periodic_offsets.size();
```
Comment (OGaboriault) [inline source/dem/periodic_boundaries_manipulator.cc:242]: why site_t and not unsigned int?

Comment (blaisb) [reply source/dem/periodic_boundaries_manipulator.cc:242]: Size t is the type used by the containers. It will compile to unsigned int, but it's acutally more robust to use size_t.

File: doc/source/parameters/dem/model_parameters.rst (hunk 2 above)
Comment (blaisb) [reply doc/source/parameters/dem/model_parameters.rst:184]: That's a good idea, but let's do that in another PR since this is unrelated :)

File: source/core/parameters_lagrangian.cc (hunk 1 above)
Comment (blaisb) [reply source/core/parameters_lagrangian.cc]: Done

File: CHANGELOG.md [hunk 5]
```diff
@@ -3,14 +3,18 @@ All notable changes to the Lethe project will be documented in this file.
...
+- MAJOR The DEM and CFD-DEM solver could only manage one set of periodic boundary condition. This PR extends the periodicity to be able to manage multiple periodic boundary conditions such that up to three pair of periodic boundary faces can be declared in 3D. [#1983](https://github.com/chaos-polymtl/lethe/pull/1983)
 
 ## [Master] - 2026/05/12
 
+### Fixed
+
+- MAJOR Fixed the usage of matrix free solver with solid material id in the domain for both the GCMG and LSMG multigrid preconditioners. This has been tested with the `grid_uniform_channel_with_meshed_cylinder` and required the mesh to have an additional option to disable the transfinite manifold has it does not work for multi-grid which have also been added to this PR. Finally, a test for the mesh `grid_uniform_channel_with_meshed_cylinder` has also been created to keep track of its changes in the future. [#1984](https://github.com/chaos-polymtl/lethe/pull/1984)
```
Comment (karlwilson) [inline CHANGELOG.md]: Corrections to this sentence:

This has been tested with the `grid_uniform_channel_with_meshed_cylinder` and required the mesh to have an additional option to disable the transfinite manifold has it does not work for multi-grid which have also been added to this PR.

should be

This has been tested with the `grid_uniform_channel_with_meshed_cylinder` and required the mesh to have an additional option to disable the transfinite manifold **as** it does not work for multi-grid which **has** also been added to this PR.

**Review by karlwilson — APPROVED** [review]
Minor comment otherwise good to go!
